### PR TITLE
feat(derive): Introduce flatten attribute

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/prometheus/client_rust/issues/117

This introduces a new attribute to the derive code for to flatten. This matches https://serde.rs/attr-flatten.html.

This seems to work, but I have some concern that if there are duplicate fields we end up encoding all of them which may make things go awry. Checking this is hard since we would need to lookup the type, which I am not sure how to do...

### CODE REVIEW

**Comment by mxinden** [comment]
> This seems to work, but I have some concern that if there are duplicate fields we end up encoding all of them which may make things go awry. Checking this is hard since we would need to lookup the type, which I am not sure how to do...

If I am not mistaken, `serde` isn't able to handle duplicate fields at compile time either.

File: derive-encode/src/lib.rs
Comment (mxinden) [inline derive-encode/src/lib.rs:29]: Today the derive macro does not support any attributes. With that in mind, what do you think of providing an error message here in case a user provides anything other than `flatten`? E.g. in the case of `flaaaaatten` it would be helpful for the user to get an error message along the lines of: "Provided attribute `flaaaaatten` but only `flatten` is supported.". What do you think?

File: derive-encode/src/lib.rs
Comment (mxinden) [inline derive-encode/src/lib.rs:32]:
```suggestion
                    let ident = f.ident.unwrap();
                    if flatten {
```

This can be pulled outside of the `if` `else`, right?

File: derive-encode/src/lib.rs
Comment (mxinden) [inline derive-encode/src/lib.rs:34]: The to-be-flattened object needs to be a `struct` and not an `enum`. In other words it needs to implement `EncodeLabelSet` and not `EncodeLabelValue`. Both traits have the same method name, namely `encode`. To prevent a misuse of the `EncodeLabelValue` trait, how about a fully qualified call here?

```suggestion
                            EncodeLabelSet::encode(self.#ident, encoder)?;
```

**Review by mxinden — COMMENTED** [review]
Great work. Thanks @howardjohn. I have a couple of questions and comments. Overall this looks good to me.

We can release this as a patch release. Would you mind bumping the patch version in `derive-encode/Cargo.toml` and provide a changelog entry in `CHANGELOG.md`?

File: derive-encode/src/lib.rs
Comment (howardjohn) [reply derive-encode/src/lib.rs]: Makes sense. Good idea, thanks!

File: CHANGELOG.md
Comment (mxinden) [inline CHANGELOG.md:14]:
```suggestion
  a nested struct to be flattened during encoding. See [PR 118].
```

File: CHANGELOG.md
Comment (mxinden) [inline CHANGELOG.md:15]: What do you think of adding your test as an example here?

``` rust
    #[derive(EncodeLabelSet, Hash, Clone, Eq, PartialEq, Debug)]
    struct CommonLabels {
        a: u64,
        b: u64,
    }
    #[derive(EncodeLabelSet, Hash, Clone, Eq, PartialEq, Debug)]
    struct Labels {
        unique: u64,
        #[prometheus(flatten)]
        common: CommonLabels,
    }

    // Would result in `my_metric{a="42",b="42",unique="42"} 42
```

I am sure folks upgrading would greatly appreciate it.

**Review by mxinden — COMMENTED** [review]
Wonderful. Thanks for the follow-up. One more thing for the changelog. Otherwise this is ready to go. Happy to cut a release right once this is merged.

File: CHANGELOG.md
Comment (howardjohn) [reply CHANGELOG.md:31]: Great idea, I updated it

**Comment by mxinden** [comment]
For legal reasons you need to add a `sign-off` line to each of your commits :(, see failing CI check. Mind doing a force push. Sorry for the inconvenience @howardjohn.

**Comment by howardjohn** [comment]
No problem, updated

**Review by mxinden — APPROVED** [review]
Thanks for the contribution! Thanks for bearing with me.

**Comment by mxinden** [comment]
[Tagged](https://github.com/prometheus/client_rust/releases/tag/prometheus-client-derive-encode-v0.4.1) and [published](https://crates.io/crates/prometheus-client-derive-encode/0.4.1).

**Pull request merged**
